### PR TITLE
optimize for the last step of go n steps plan

### PR DESCRIPTION
**Pull request by jievince**
## What type of PR is this?
- [ ] bug
- [ ] feature
- [x] enhancement

#### Description:
Optimize the last step of go n steps plan.
In the following query, the where clause only uses dst tag props, so we still can replace the last `Getneightbors` with `GetDstbySrc`.
```
GO 3 STEPS FROM "Tony Parker" OVER serve BIDIRECT WHERE $$.team.name != "Lakers" YIELD DISTINCT id($$)
```

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

**Comment by Shylock-Hg** [comment]
Can we do it in optimizer?

**Comment by jievince** [comment]
> Can we do it in optimizer?

It's very hard to do it in the optimizer.

**Review by nevermore3 — APPROVED** [review]

File: src/graph/planner/ngql/GoPlanner.cpp [hunk 1]
```diff
@@ -626,18 +640,31 @@ StatusOr<SubPlan> GoPlanner::transform(AstContext* astCtx) {
...
   if (!exprProps.dstTagProps().empty()) return false;
-  for (auto& edgeProp : exprProps.edgeProps()) {
+
+  return goCtx_->edgePropsOnlyUseDstId;
+}
+
+bool GoPlanner::edgePropsOnlyUseDstId() {
+  for (auto& edgeProp : goCtx_->exprProps.edgeProps()) {
     auto props = edgeProp.second;
     if (props.size() != 1) return false;
```
Comment (czpmango) [inline src/graph/planner/ngql/GoPlanner.cpp:659]: Don't check size, bcz some scenarios that can be optimized may be ignored.

File: src/graph/planner/ngql/GoPlanner.cpp [hunk 2]
```diff
@@ -626,18 +640,31 @@ StatusOr<SubPlan> GoPlanner::transform(AstContext* astCtx) {
...
     if (props.find(kDst) == props.end()) return false;
   }
 
+  return true;
+}
+
+bool GoPlanner::onlyYieldDistinctDstId() {
+  if (!goCtx_->distinct) return false;
   if (goCtx_->yieldExpr->columns().size() != 1) {
     return false;
```
Comment (czpmango) [inline src/graph/planner/ngql/GoPlanner.cpp:669]: Foreach yield items and check if expression contains only constant or id expressions.

File: src/graph/planner/ngql/GoPlanner.cpp [hunk 3]
```diff
@@ -626,18 +640,31 @@ StatusOr<SubPlan> GoPlanner::transform(AstContext* astCtx) {
...
     if (props.size() != 1) return false;
     if (props.find(kDst) == props.end()) return false;
   }
 
+  return true;
+}
+
+bool GoPlanner::onlyYieldDistinctDstId() {
+  if (!goCtx_->distinct) return false;
   if (goCtx_->yieldExpr->columns().size() != 1) {
```
Comment (czpmango) [inline src/graph/planner/ngql/GoPlanner.cpp:668]: Ditto. Dont check size.

File: src/graph/planner/ngql/GoPlanner.cpp (hunk 2 above)
Comment (czpmango) [reply src/graph/planner/ngql/GoPlanner.cpp:669]: yield distinct like._dst, like._dst,8,7

**Review by czpmango — APPROVED** [review]

File: src/graph/planner/ngql/GoPlanner.cpp [hunk 4]
```diff
@@ -310,41 +310,53 @@ PlanNode* GoPlanner::buildLastStepJoinPlan(PlanNode* gn, PlanNode* join) {
...
+
+  if (!goCtx_->joinInput && goCtx_->limits.empty() && goCtx_->onlyYieldDistinctDstId &&
+      !goCtx_->exprProps.hasSrcTagProperty() && goCtx_->edgePropsOnlyUseDstId) {
     auto* gd = GetDstBySrc::make(qctx, dep, goCtx_->space.id);
     gd->setSrc(goCtx_->from.src);
     gd->setEdgeTypes(buildEdgeTypes());
     gd->setInputVar(goCtx_->vidsVar);
-    gd->setColNames({kDst});
-    scan = gd;
+    gd->setColNames(goCtx_->colNames);
```
[the remark this answers is not in the document]
Comment (jievince) [reply src/graph/planner/ngql/GoPlanner.cpp:320]: Yes, it's ok.

File: src/graph/context/ast/QueryAstContext.h [hunk 5]
```diff
@@ -95,6 +95,8 @@ struct GoContext final : AstContext {
   bool joinDst{false};
   // Optimize for some simple go sentence which only need dst id.
   // eg. GO 1 TO N STEPS FROM "A" OVER like YIELD DISTINCT like._dst
+  bool onlyYieldDistinctDstId{false};
+  bool edgePropsOnlyUseDstId{false};
   bool isSimple{false};
```
[the remark this answers is not in the document]
Comment (jievince) [reply src/graph/context/ast/QueryAstContext.h:100]: It's useful. It's used by MToNSteps.

**Review by nevermore3 — APPROVED** [review]

**Pull request merged**